internal/k8s/controller.go

```diff
@@ -792,6 +792,11 @@ func (lbc *LoadBalancerController) syncPolicy(task task) {
 		return
 	}
 
+	if !lbc.HasCorrectIngressClass(obj) {
```

Comment:
Unfortunately, we can't simply ignore a policy here. Consider a case if the policy IngressClassName is updated:
a) IngressClassName is updated to the one that doesn't match the class of the IC. In this case, the IC needs to regenerate the configuration for all affected VSes 
b) IngressClassName is updated to the one that matches the class of the IC. In this case, the IC needs also to regenerate the configuration for all affected VSes 


I think adding the IngressClass check in `getPolicies()` will solve the problem - if the Policy IngressClass doesn't match, getPolicies returns an error. As a result, the IC will generate the correct config (with `return 500`) and report the error in the VS/VSR status/events.

Note that we will still need to keep not reporting status/events for policies of a different IngressClass.

Reply:
updated to use `getPolicies()`

Reply:
> Note that we will still need to keep not reporting status/events for policies of a different IngressClass.


https://github.com/nginxinc/kubernetes-ingress/blob/f1ebca1e3524346c6840e9256be208f2638f525e/internal/k8s/controller.go#L797-L821

because an IC can report status for a Policy of a different IngressClass, this could cause problems. Consider an case of two ICs:
(1) with class `one` with enabled preview policies
(2) twi class `two` with disabled preview policies. 

If a user creates a preview policy with IngressClassName `one`, the (2) IC will still update the status of the Policy. If that happens after the (1) IC updated the status, the user will see a confusing error (the (2) IC will update the status with an error because preview policies are disabled) 

Could you make sure the IC only reports status for policies of its IngressClass?

Reply:
updated

Reply:
Hi @soneillf5  could you also disable reporting events for policies of a different class? sorry, I mentioned the events in my first comment, but didn't mention them in the second one.

Reply:
Can you clarify this? I was looking at the status updating logic and didn't see any other exclusion logic for TransportServer or VirtualServer. It's very painful to develop a feature via pull request.

Reply:
Sure. VirtualServer and TransportServer are processed through the Configuration type.  The Configuration type checks for the resource class and if it doesn't match - it will ignore it -- https://github.com/nginxinc/kubernetes-ingress/blob/master/internal/k8s/configuration.go#L465  At the end, no change or problem for it will be generated so no status will be reported or event emitted.

In contrast, the Policy resource isn't processes through the Configuration type, so that's why we need to add that check.  If you condition this code https://github.com/nginxinc/kubernetes-ingress/blob/dfad4850b02cedc3b08deb72643677b475eee91d/internal/k8s/controller.go#L798-L822 so that it doesn't get invoked when the policy is of a different class, that will solve the problem of both reporting an event and the status for the policy of a different class.

Reply:
> If you condition this code ... so that it doesn't get invoked when the policy is of a different class
I'm not following you again sorry, you mean add a check for the ingress class and if it doesn't match, don't invoke this code?

Reply:
Out of interest, I've updated the tests to capture the change around the status. The test no confirms that if the ingress class does not match, the status of the policy is not updated. `tests/suite/test_policy_ingress_class.py:143`

Reply:
> if polExists && !hasCorrectIngressClass(obj) {

👍 

>  The test no confirms that if the ingress class does not match, the status of the policy is not updated.

what about the event for the Policy resource?

Reply:
sorted